### ISSUE TITLE
add NVTX markers to TimeProfile

### ISSUE THREAD

**Issue by mathiaswagner** [lattice/quda#298]
QUDA uses timers to measure time spent in specific parts.
To simplify visual profiling we could add NVTX markers in the struct `TimeProfile`
```c++
  struct TimeProfile {
    std::string fname;  /**< Which function are we profiling */

    Timer profile[QUDA_PROFILE_COUNT];
    static std::string pname[];

    bool switchOff;
    
    TimeProfile(std::string fname) : fname(fname), switchOff(false) { ; }

    /**< Print out the profile information */
    void Print();

    void Start(QudaProfileType idx) { 
      // if total timer isn't running, then start it running
      if (!profile[QUDA_PROFILE_TOTAL].running && idx != QUDA_PROFILE_TOTAL) {
	profile[QUDA_PROFILE_TOTAL].Start(); 
	switchOff = true;
      }

      profile[idx].Start(); 
    }

    void Stop(QudaProfileType idx) { 
      profile[idx].Stop(); 

      // switch off total timer if we need to
      if (switchOff && idx != QUDA_PROFILE_TOTAL) {
	profile[QUDA_PROFILE_TOTAL].Stop(); 
	switchOff = false;
      }
    }

    double Last(QudaProfileType idx) { 
      return profile[idx].last;
    }

  };
```

For a first implementation it is probably sufficient to use the NVTX markup code from Jiri ( http://devblogs.nvidia.com/parallelforall/cuda-pro-tip-generate-custom-application-profile-timelines-nvtx/ )and add `PUSH_RANGE(fname,idx)` to the `Start` call and `POP_RANGE` to the `Stop` call.
`name` already gives the function name and to distinguish between compute, transfers, ... we could map the enum in `QudaProfileType` to different colors.

Using a configure option this would generate zero overhead in production code.

We already do have NVTX markup for MPI calls and I have started adding support for that in the MILC interface #262

**Comment by mikeaclark** [comment]
This would be great!  How do multiple concurrent NVTX markers coexist?  E.g., it's possible that we have both MPI/ MILC interface and Profile NVTX all pushed simultaneously.  I guess the visual profiler does a nice job displaying this?

**Comment by mathiaswagner** [comment]
I have not tried with multiple (overlapping) markers yet.
But the screen shot at the end of the section *Using compiler instrumentation with NVTX* in Jiri's post looks like that will work.

I'll give it a try.

**Comment by mathiaswagner** [comment]
I checked back because of the license and it is covered by: https://github.com/parallel-forall/code-samples/blob/master/license.txt

So we should probably include that license file.

**Comment by mikeaclark** [comment]
Are there any implications of dual BSD /Apache licensing?

**Comment by mathiaswagner** [comment]
As far as I can tell BSD and Apache licenses are similar. Apache requires the inclusion of the license, state changes and include the notice (if exists). As far as I can tell it only applies to the file included or probably the code included from it. 
Maybe @rbabich has an understanding of it.
Jiri did not specifically apply a license to the code but it is included in the parallel forall repository ...
Even if, I think the code fragment we use would be covered under FAIR use. A lot of the code is basically identical to the NVTX code documentation.

**Comment by rbabich** [comment]
Apache includes a patent grant, so it's substantially different from MIT or BSD.  (QUDA is currently licensed under MIT.)

Kate: Maybe just ask Jiri if he minds us including snippets of his code under our current license?  (Technically, he'd be joining the ranks of "QUDA Developers" mentioned in the LICENSE file.)

**Comment by mikeaclark** [comment]
Isn't QUDA BSD, not MIT?  At least it only mentions BSD in the license not MIT.....

(There's already a side discussion between Jiri and Mark H. as to how to license this.)

**Comment by rbabich** [comment]
It's MIT.  The "additional BSD notice" is referring to the license for LLNL's PMPI wrapper.  There should probably be better indentation or something to make that clear.

**Comment by mathiaswagner** [comment]
Jiri has added a BSD license to the NVTX code:

```
/* Copyright (c) 2014, NVIDIA CORPORATION. All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of NVIDIA CORPORATION nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
```

If I get it right all we need to do is to include that in our LICENSE file and we can merge that in?

**Comment by rbabich** [comment]
Yep.  I'd preface it with a mention of the specific file names that contain the code.  Then remove the comment delimiters and maybe indent it a few spaces.  (It would also be a good idea to block-indent the licenses for gtest and PMPI.)

**Comment by mikeaclark** [comment]
closed by #342
